java-checks/src/main/java/org/sonar/java/checks/TooManyStatementsPerLine_S00122_Check.java

```diff
@@ -47,91 +74,231 @@
+    super.scanFile(context);
   }
 
   @Override
-  public void visitFile(AstNode astNode) {
-    statementsPerLine.clear();
-    columnsByLine.clear();
+  public void visitNode(Tree tree) {
+    BlockTree block;
+    if (tree.is(Tree.Kind.STATIC_INITIALIZER)) {
+      block = (BlockTree) tree;
+    } else {
+      block = ((MethodTree) tree).block();
+    }
+    if (block != null) {
+      StatementVisitor visitor = new StatementVisitor();
+      block.accept(visitor);
+      for (Entry<Integer> entry : visitor.statementsPerLine.entrySet()) {
+        int count = entry.getCount();
+        if (count > 1) {
+          addIssue(entry.getElement(), "At most one statement is allowed per line, but " + count + " statements were found on this line.");
+        }
+      }
+    }
   }
 
-  @Override
-  public void visitNode(AstNode statementNode) {
-    if (statementNode.is(JavaLexer.VARIABLE_DECLARATORS) && !statementNode.hasParent(JavaLexer.BLOCK_STATEMENTS)) {
-      return;
+  private class StatementVisitor extends BaseTreeVisitor {
+    private final Multiset<Integer> statementsPerLine = HashMultiset.create();
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      // do nothing, as inner class will be visited later
+    }
+
+    @Override
+    public void visitAssertStatement(AssertStatementTree tree) {
+      addLines(tree.assertKeyword(), tree.semicolonToken());
+    }
+
+    @Override
+    public void visitBreakStatement(BreakStatementTree tree) {
+      addLines(tree.breakKeyword(), tree.semicolonToken());
     }
 
-    if (!isExcluded(statementNode)) {
-      int lineStart = statementNode.getTokenLine();
-      int lineEnd = statementNode.getLastToken().getLine();
-      int columnStart = statementNode.getToken().getColumn();
-      int columnEnd = statementNode.getLastToken().getColumn();
+    @Override
+    public void visitContinueStatement(ContinueStatementTree tree) {
+      addLines(tree.continueKeyword(), tree.semicolonToken());
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree tree) {
+      addLines(tree.returnKeyword(), tree.semicolonToken());
+    }
 
-      if (!isNestedInStatement(lineStart, columnStart)) {
-        statementsPerLine.add(lineStart);
+    @Override
+    public void visitThrowStatement(ThrowStatementTree tree) {
+      addLines(tree.throwKeyword(), tree.semicolonToken());
+    }
+
+    @Override
+    public void visitExpressionStatement(ExpressionStatementTree tree) {
+      SyntaxToken firstToken = getFirstToken(tree.expression());
+      SyntaxToken endToken = tree.semicolonToken();
+      if (firstToken != null) {
+        addLines(firstToken, endToken);
+      } else {
+        addLine(endToken);
       }
-      if (lineStart != lineEnd) {
-        if (!isNestedInStatement(lineEnd, columnEnd)) {
-          statementsPerLine.add(lineEnd);
+    }
+
+    @CheckForNull
+    private SyntaxToken getFirstToken(Tree expression) {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)The Cyclomatic Complexity of this method "getFirstToken" is 21 which is greater than 10 authorized.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Reduce the number of returns of this method 11, down to the maximum allowed 5.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1142)